libp2p/stream_muxer/mplex/mplex_stream.py

```diff
@@ -58,19 +58,41 @@ def is_initiator(self) -> bool:
         done, pending = await asyncio.wait(  # type: ignore
             [
                 self.event_reset.wait(),
-                self.event_remote_closed.wait(),
                 self.incoming_data.get(),
+                self.event_remote_closed.wait(),
             ],
             return_when=asyncio.FIRST_COMPLETED,
         )
+        for fut in pending:
+            fut.cancel()
+
         if self.event_reset.is_set():
             raise MplexStreamReset
+
+        if len(done) != 1:
+            raise Exception(f"Should be exactly 1 job in {done}.")
+        done_task = tuple(done)[0]
+        # NOTE: Ignore type check because the typeshed for `asyncio.Task` does not
+        #   have the field `_coro`.
+        coro_qualname = done_task._coro.__qualname__  # type: ignore
+        # If `qualname == "Queue.get"` then there is incoming data. We can add it to the buffer.
+        if coro_qualname == "Queue.get":
```

Comment:
we probably want a more stable way to do this moving forward (e.g. what happens if they rename `Queue.get` in python 3.8, etc...)

but if it works now, i think its good enough to merge in so we can leverage this in interop -- we have been seeing some stuff around not closing streams across the clients

Reply:
I changed the code `_wait_for_data` to compare the tasks precisely, instead of checking the function name. This should not be an issue anymore. Thanks a lot for bringing this up!